File: app/reasoning/log_parser.py
"""Heuristic detection and parsing of error logs / stack traces."""
import re
from dataclasses import dataclass, field
from typing import List


@dataclass
class LogSignature:
    error_codes: List[str] = field(default_factory=list)
    exception_types: List[str] = field(default_factory=list)
    stack_trace_lines: List[str] = field(default_factory=list)
    error_messages: List[str] = field(default_factory=list)
    is_log: bool = False


# ---------------------------------------------------------------------------
# Detection patterns
# ---------------------------------------------------------------------------

_PYTHON_TB = re.compile(r"Traceback \(most recent call last\)", re.IGNORECASE)
_JAVA_FRAME = re.compile(r"^\s+at\s+[\w.$<>]+\(", re.MULTILINE)
_JS_FRAME = re.compile(r"^\s+at\s+.+:\d+:\d+", re.MULTILINE)
_EXCEPTION_LINE = re.compile(
    r"(Exception|Error|FATAL|CRITICAL|Traceback|Caused by)[\s:]",
    re.IGNORECASE,
)
_HTTP_5XX = re.compile(r"\bHTTP[/ ]+5\d{2}\b", re.IGNORECASE)
_HTTP_4XX = re.compile(r"\bHTTP[/ ]+4\d{2}\b", re.IGNORECASE)
_ERROR_CODE = re.compile(
    r"\b(E\d{4,}|ORA-\d+|SQLSTATE\[\d+\]|errno\s*\d+|code\s*[:=]\s*\d{3,})\b",
    re.IGNORECASE,
)
_LOG_TIMESTAMP = re.compile(
    r"(\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}|\[\d{4}-\d{2}-\d{2})",
    re.IGNORECASE,
)
_PIPE_BRACKET_LINE = re.compile(r"[\[\|]{1}.{10,}[\]\|]{1}")


def is_error_log(text: str) -> bool:
    """Detect whether *text* looks like an error log or stack trace.

    Heuristics (any 1 strong signal OR 2+ weak signals triggers True):
    - Python Traceback header
    - Java/JS stack frame lines (at com.example...)
    - Exception:/Error:/FATAL:/CRITICAL: keywords
    - HTTP 5xx error codes
    - Oracle/DB error codes (ORA-XXXXX, E0XXXX)
    - Log timestamps on multiple lines
    - Log-style pipe/bracket patterns on multiple lines
    """
    strong = [
        bool(_PYTHON_TB.search(text)),
        bool(_JAVA_FRAME.search(text)),
        bool(_JS_FRAME.search(text)),
    ]
    if any(strong):
        return True

    weak_score = 0
    weak_score += 1 if _EXCEPTION_LINE.search(text) else 0
    weak_score += 1 if _HTTP_5XX.search(text) else 0
    weak_score += 1 if _ERROR_CODE.search(text) else 0

    ts_matches = _LOG_TIMESTAMP.findall(text)
    weak_score += 1 if len(ts_matches) >= 2 else 0

    bracket_matches = _PIPE_BRACKET_LINE.findall(text)
    weak_score += 1 if len(bracket_matches) >= 2 else 0

    return weak_score >= 2


# ---------------------------------------------------------------------------
# Signature extraction
# ---------------------------------------------------------------------------

_EXCEPTION_TYPE_RE = re.compile(
    r"^(?:\s+Caused by:\s+)?([A-Za-z][\w.]*(?:Exception|Error|Fault|Panic))\b",
    re.MULTILINE,
)
_STACK_FRAME_RE = re.compile(r"^\s+(?:at\s+.+|File\s+\".+\",\s+line\s+\d+)", re.MULTILINE)
_ERROR_MSG_RE = re.compile(
    r"(Exception|Error|FATAL|CRITICAL)[:\s]+(.{10,120})",
    re.IGNORECASE,
)


def parse_log_signature(text: str) -> LogSignature:
    """Extract a structured signature from an error log for improved search."""
    sig = LogSignature(is_log=is_error_log(text))

    # Error codes
    sig.error_codes = list({m.group(0) for m in _ERROR_CODE.finditer(text)})

    # HTTP codes
    http_matches = re.findall(r"\bHTTP[/ ]+(\d{3})\b", text, re.IGNORECASE)
    sig.error_codes.extend([f"HTTP_{c}" for c in http_matches])
    sig.error_codes = list(dict.fromkeys(sig.error_codes))  # dedup preserving order

    # Exception types
    sig.exception_types = list(
        dict.fromkeys(m.group(1) for m in _EXCEPTION_TYPE_RE.finditer(text))
    )

    # Stack trace lines
    sig.stack_trace_lines = _STACK_FRAME_RE.findall(text)[:10]

    # Error messages
    sig.error_messages = [
        m.group(2).strip() for m in _ERROR_MSG_RE.finditer(text)
    ][:5]

    return sig


def build_search_query(original_query: str, sig: LogSignature) -> str:
    """Build an enhanced search query from a log signature."""
    terms: List[str] = []
    terms.extend(sig.exception_types[:3])
    terms.extend(sig.error_codes[:3])
    terms.extend(sig.error_messages[:2])
    if terms:
        return " ".join(terms)
    return original_query

File: Merlin.CSharp/src/Merlin/Reasoning/LogParser.cs
using System.Text.RegularExpressions;
using Merlin.Models;

namespace Merlin.Reasoning;

/// <summary>
/// Heuristic detection and parsing of error logs / stack traces.
/// Mirrors app/reasoning/log_parser.py.
/// </summary>
public static class LogParser
{
    // ── Detection patterns ────────────────────────────────────────────────────

    private static readonly Regex PythonTb = new(
        @"Traceback \(most recent call last\)", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex JavaFrame = new(
        @"^\s+at\s+[\w.$<>]+\(", RegexOptions.Multiline | RegexOptions.Compiled);

    private static readonly Regex JsFrame = new(
        @"^\s+at\s+.+:\d+:\d+", RegexOptions.Multiline | RegexOptions.Compiled);

    private static readonly Regex ExceptionLine = new(
        @"(Exception|Error|FATAL|CRITICAL|Traceback|Caused by)[\s:]",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex Http5Xx = new(
        @"\bHTTP[/ ]+5\d{2}\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex ErrorCode = new(
        @"\b(E\d{4,}|ORA-\d+|SQLSTATE\[\d+\]|errno\s*\d+|code\s*[:=]\s*\d{3,})\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex LogTimestamp = new(
        @"(\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}|\[\d{4}-\d{2}-\d{2})",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex PipeBracketLine = new(
        @"[\[\|]{1}.{10,}[\]\|]{1}", RegexOptions.Compiled);

    /// <summary>
    /// Detect whether text looks like an error log or stack trace.
    /// Heuristics: any 1 strong signal OR 2+ weak signals triggers true.
    /// </summary>
    public static bool IsErrorLog(string text)
    {
        // Strong signals
        if (PythonTb.IsMatch(text) || JavaFrame.IsMatch(text) || JsFrame.IsMatch(text))
            return true;

        // Weak signals
        int weakScore = 0;
        if (ExceptionLine.IsMatch(text)) weakScore++;
        if (Http5Xx.IsMatch(text))       weakScore++;
        if (ErrorCode.IsMatch(text))     weakScore++;
        if (LogTimestamp.Matches(text).Count >= 2) weakScore++;
        if (PipeBracketLine.Matches(text).Count >= 2) weakScore++;

        return weakScore >= 2;
    }

    // ── Signature extraction patterns ─────────────────────────────────────────

    private static readonly Regex ExceptionTypeRe = new(
        @"^(?:\s+Caused by:\s+)?([A-Za-z][\w.]*(?:Exception|Error|Fault|Panic))\b",
        RegexOptions.Multiline | RegexOptions.Compiled);

    private static readonly Regex StackFrameRe = new(
        @"^\s+(?:at\s+.+|File\s+"".+"",\s+line\s+\d+)",
        RegexOptions.Multiline | RegexOptions.Compiled);

    private static readonly Regex ErrorMsgRe = new(
        @"(Exception|Error|FATAL|CRITICAL)[:\s]+(.{10,120})",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex HttpCodeRe = new(
        @"\bHTTP[/ ]+(\d{3})\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>Extract a structured signature from an error log for improved search.</summary>
    public static LogSignature ParseLogSignature(string text)
    {
        var sig = new LogSignature { IsLog = IsErrorLog(text) };

        // Error codes (DB/application)
        sig.ErrorCodes = ErrorCode.Matches(text)
            .Select(m => m.Value)
            .Distinct()
            .ToList();

        // HTTP codes
        foreach (Match m in HttpCodeRe.Matches(text))
            if (!sig.ErrorCodes.Contains($"HTTP_{m.Groups[1].Value}"))
                sig.ErrorCodes.Add($"HTTP_{m.Groups[1].Value}");

        // Exception types (dedup preserving order)
        sig.ExceptionTypes = ExceptionTypeRe.Matches(text)
            .Select(m => m.Groups[1].Value)
            .Distinct()
            .ToList();

        // Stack trace lines (max 10)
        sig.StackTraceLines = StackFrameRe.Matches(text)
            .Select(m => m.Value)
            .Take(10)
            .ToList();

        // Error messages (max 5)
        sig.ErrorMessages = ErrorMsgRe.Matches(text)
            .Select(m => m.Groups[2].Value.Trim())
            .Take(5)
            .ToList();

        return sig;
    }

    /// <summary>Build an enhanced search query from a log signature.</summary>
    public static string BuildSearchQuery(string originalQuery, LogSignature sig)
    {
        var terms = new List<string>();
        terms.AddRange(sig.ExceptionTypes.Take(3));
        terms.AddRange(sig.ErrorCodes.Take(3));
        terms.AddRange(sig.ErrorMessages.Take(2));

        return terms.Count > 0 ? string.Join(" ", terms) : originalQuery;
    }
}
